frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx

```diff
@@ -0,0 +1,332 @@
+        RecordPlugin.getAvailableAudioDevices().then(devices => {
+          setAvailableAudioDevices(devices)
+          if (devices.length > 0) {
+            setActiveAudioDeviceId(devices[0].deviceId)
+          }
+        })
+      })
+      .catch(_err => {
+        setHasNoMicPermissions(true)
+      })
+  }, [])
+
+  const initializeWaveSurfer = useCallback(() => {
+    if (waveSurferRef.current === null) return
+
+    const ws = WaveSurfer.create({
+      container: waveSurferRef.current,
+      waveColor: theme.colors.primary,
+      progressColor: theme.colors.bodyText,
+      height: WAVEFORM_HEIGHT - 2 * WAVEFORM_PADDING,
+      barWidth: BAR_WIDTH,
+      barGap: BAR_GAP,
+      barRadius: BAR_RADIUS,
+      cursorWidth: CURSOR_WIDTH,
+    })
+
+    ws.on("timeupdate", time => {
+      setProgressTime(formatTime(time * 1000)) // get from seconds to milliseconds
+    })
+
+    ws.on("pause", () => {
+      forceRerender()
+    })
+
+    const recordPlugin = ws.registerPlugin(
+      RecordPlugin.create({
+        scrollingWaveform: false,
+        renderRecordedAudio: true,
+      })
+    )
+
+    recordPlugin.on("record-end", blob => {
+      const url = URL.createObjectURL(blob)
+      setRecordingUrl(url)
+
+      const file = new File([blob], "audio.wav", { type: blob.type })
+      uploadTheFile(file)
+
+      ws.setOptions({
+        // This color is hardcoded because I've spent an hour trying to get it to work
+        // with the theme and I'm giving up for now. The complications arise due to the
+        // way wavesurfer is using these colors. Specifically, the progress color "tints" the wave color
+        // and the opacity of this wave color is not being respected, so to make things work for now hardcoding
+        // this color that looks "okay" on both themes.
+        waveColor: "#A5A5AA",
+      })
+    })
+
+    recordPlugin.on("record-progress", time => {
+      setRecordingTime(formatTime(time))
+    })
+
+    setWavesurfer(ws)
+    setRecordPlugin(recordPlugin)
+
+    return () => {
+      if (ws) ws.destroy()
+    }
+  }, [theme, uploadTheFile])
+
+  useEffect(() => {
+    initializeWaveSurfer()
+  }, [initializeWaveSurfer])
+
+  const onClickPlayPause = useCallback(() => {
+    if (wavesurfer) {
+      wavesurfer.playPause()
+      // This is because we want the time to be the duration of the audio when they stop recording,
+      // but once they start playing it, we want it to be the current time. So, once they start playing it
+      // we'll start keeping track of the playback time from that point onwards (until re-recording).
+      setShouldUpdatePlaybackTime(true)
+
+      // to get the pause button to show
+      forceRerender()
+    }
+  }, [wavesurfer])
+
+  const startRecording = useCallback(() => {
+    if (!recordPlugin || !activeAudioDeviceId || !wavesurfer) {
+      return
+    }
+
+    const deviceId = activeAudioDeviceId
+    if (isNullOrUndefined(deviceId)) {
+      return
+    }
```

Comment:
isn't this already checked via the check in [line 209](https://github.com/streamlit/streamlit/pull/9404/files#diff-a7eb8abbe0fa76f9716f9f399ee823661f1a6c761fa6065ef32204746da10ad4R209) `!activeAudioDeviceId`?

Reply:
yup, fixed